### PR TITLE
hw10-observability submission

### PR DESCRIPTION
**Pull request by vladyslavpavlenko**
A detailed overview of this pull request, as well as the comments on alerts, can be found in [hw10-observability.md](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-vladyslavpavlenko/blob/dc49a000ed78c26bf9dae8b8e5eafe55749e5cba/docs/observability_hw10.md).

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+.git
+.gitignore
+Dockerfile
+docker-compose.yml
+node_modules
```
Comment (vladpistun) [inline .dockerignore]: u have js packages here?

File: cmd/api/main.go [hunk 2]
```diff
@@ -1,19 +1,17 @@
...
 	"github.com/vladyslavpavlenko/genesis-api-project/internal/app/config"
+	"github.com/vladyslavpavlenko/genesis-api-project/pkg/logger"
 )
 
 func main() {
-	appConfig := config.NewAppConfig()
+	a := config.New()
+	a.Logger = logger.New()
 
-	err := app.Run(appConfig)
```
Comment (vladpistun) [inline cmd/api/main.go:14]: better devide config and logger

File: cmd/api/main.go [hunk 3]
```diff
@@ -1,19 +1,17 @@
...
+	err := app.Run(a)
 	if err != nil {
-		log.Printf("Error: %v\n", err)
-		os.Exit(1)
+		a.Logger.Fatal(err.Error())
```
Comment (vladpistun) [inline cmd/api/main.go]: this log wont be called

File: internal/app/run.go [hunk 4]
```diff
@@ -147,19 +150,19 @@ func scheduleEmails(s scheduler, n *notifier.Notifier) error {
...
 
 // eventConsumer runs an event dispatcher.
-func eventConsumer(ctx context.Context, c consumer) {
+func eventConsumer(ctx context.Context, c consumer, log *logger.Logger) {
 	go c.Consume(ctx)
 
 	// Wait for context cancellation to handle graceful shutdown
 	<-ctx.Done()
-	log.Println("Shutting down event consumer...")
+	log.Info("shutting down event consumer...")
```
Comment (vladpistun) [inline internal/app/run.go]: better use l, applogger -> dont use name as  default go package name

File: internal/email/consumer/kafka_consumer.go [hunk 5]
```diff
@@ -46,24 +51,40 @@ func NewKafkaConsumer(kafkaURL, topic string, partition int, groupID string, sen
...
 }
 
 // Consume is a worker that consumes messages from Kafka and processes them
 // to send an email using the Sender interface.
 func (c *KafkaConsumer) Consume(ctx context.Context) {
 	for {
+		// Check if context is canceled before attempting to fetch a message
+		select {
+		case <-ctx.Done():
+			c.logger.Warn("shutting down consumer...", zap.String("cause", "context canceled"))
```
Comment (vladpistun) [inline internal/email/consumer/kafka_consumer.go]:
```suggestion
			c.logger.Info("shutting down consumer...", zap.String("cause", "context canceled"))
```

File: internal/email/consumer/kafka_consumer.go [hunk 6]
```diff
@@ -46,24 +51,40 @@ func NewKafkaConsumer(kafkaURL, topic string, partition int, groupID string, sen
...
+			return
+		default:
+		}
+
 		// Attempt to fetch a message from Kafka
 		m, err := c.Reader.FetchMessage(ctx)
 		if err != nil {
-			log.Printf("Failed to read message: %v", err)
+			if errors.Is(err, context.Canceled) {
+				c.logger.Warn("shutting down consumer...", zap.String("cause", "context canceled"))
```
Comment (vladpistun) [inline internal/email/consumer/kafka_consumer.go]:
```suggestion
				c.logger.Info("shutting down consumer...", zap.String("cause", "context canceled"))
```

File: internal/email/consumer/kafka_consumer.go [hunk 7]
```diff
@@ -74,27 +95,32 @@ func (c *KafkaConsumer) Consume(ctx context.Context) {
...
 			continue
 		}
 
 		// Commit the offset back to Kafka to mark the message as processed
 		if err = c.Reader.CommitMessages(ctx, m); err != nil {
-			log.Printf("Failed to commit message offset %d: %v", m.Offset, err)
+			c.logger.Error("failed to commit message", zap.Int64("offset", m.Offset), zap.Error(err))
 		} else {
-			log.Printf("Offset committed successfully for message at offset: %d", m.Offset)
+			c.logger.Error("offset committed", zap.Int64("offset", m.Offset))
```
Comment (vladpistun) [inline internal/email/consumer/kafka_consumer.go]: why error? its successful flow

File: internal/email/consumer/kafka_consumer.go [hunk 8]
```diff
@@ -106,10 +132,10 @@ func (c *KafkaConsumer) sendMessage(data outbox.Data) {
 		Body:    fmt.Sprintf("The current exchange rate for USD to UAH is %.2f.", data.Rate),
 	}
 
-	log.Printf("Sending email to %s", data.Email)
+	c.logger.Debug("sending email", zap.String("email", data.Email))
```
Comment (vladpistun) [inline internal/email/consumer/kafka_consumer.go]: i prefer this as info, cauz we can use this logs for some investigations

File: internal/outbox/producer/kafka_producer.go [hunk 9]
```diff
@@ -94,15 +101,14 @@ func (p *KafkaProducer) processEvents(ctx context.Context, topic string, partiti
...
+		p.logger.Error("failed to start transaction", zap.Error(err))
 		return
 	}
-	log.Println("Transaction started successfully")
 
 	defer func() {
 		if r := recover(); r != nil {
 			tx.Rollback()
-			log.Printf("Recovered from panic: %v, transaction rolled back", r)
+			p.logger.Debugf("recovered from panic: %v, transaction rolled back", r)
```
Comment (vladpistun) [inline internal/outbox/producer/kafka_producer.go]: panic recover should be error, panic btw is error

File: internal/outbox/producer/kafka_producer.go [hunk 10]
```diff
@@ -142,28 +152,28 @@ func (p *KafkaProducer) processEvents(ctx context.Context, topic string, partiti
 			Partition: partition,
 		}
 
-		log.Printf("Preparing to send message with ID: %d", event.ID)
+		p.logger.Debug("sending message to kafka", zap.Int("message_id", int(event.ID)))
```
Comment (vladpistun) [inline internal/outbox/producer/kafka_producer.go]: info better

File: internal/outbox/producer/kafka_producer.go [hunk 11]
```diff
@@ -142,28 +152,28 @@ func (p *KafkaProducer) processEvents(ctx context.Context, topic string, partiti
...
 	}
 
 	// Commit the transaction if all events are processed successfully
 	if err = tx.Commit().Error; err != nil {
-		log.Printf("Failed to commit transaction: %v", err)
+		p.logger.Warn("transaction failed", zap.Error(err))
 		return
 	}
-	log.Println("Transaction committed and all events processed successfully")
+	p.logger.Debug("all events processed")
```
Comment (vladpistun) [inline internal/outbox/producer/kafka_producer.go]: what about more context, what events

File: internal/rateapi/logger.go [hunk 12]
```diff
@@ -2,36 +2,39 @@ package rateapi
...
-	}
+	"github.com/vladyslavpavlenko/genesis-api-project/pkg/logger"
 
-	Fetcher interface {
-		Fetch(ctx context.Context, base, target string) (string, error)
-	}
+	"go.uber.org/zap"
 )
 
+type Fetcher interface {
```
Comment (vladpistun) [inline internal/rateapi/logger.go]: why this interface is public?

File: internal/storage/gormstorage/driver.go [hunk 13]
```diff
@@ -3,51 +3,59 @@ package gormstorage
...
-func (c *Connection) Setup(dsn string) error {
+// Setup sets up a new Connection with a logger.
+func (c *Connection) Setup(dsn string, l *logger.Logger) error {
+	c.logger = l
 	var counts int64
 	for {
 		db, err := openDB(dsn)
 		if err != nil {
-			log.Printf("Postgres not yet ready... Attempt: %d\n", counts)
+			c.logger.Warn("Postgres not yet ready...", zap.Int64("attempt", counts), zap.Error(err))
```
Comment (vladpistun) [inline internal/storage/gormstorage/driver.go]: error should be error level

File: pkg/logger/logger.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+	l.sugaredLogger.Warnf(template, args...)
+}
+
+// Errorf logs a message at Error level with a template.
+func (l *Logger) Errorf(template string, args ...any) {
+	l.sugaredLogger.Errorf(template, args...)
+}
+
+// DPanicf logs a message at DPanic level with a template.
+func (l *Logger) DPanicf(template string, args ...any) {
```
Comment (vladpistun) [inline pkg/logger/logger.go]: prefer remove it

File: pkg/logger/logger.go [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+	l.sugaredLogger.DPanicf(template, args...)
+}
+
+// Panicf logs a message at Panic level with a template.
+func (l *Logger) Panicf(template string, args ...any) {
```
Comment (vladpistun) [inline pkg/logger/logger.go]: prefer remove it

File: pkg/logger/logger.go [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+	l.logger.Fatal(msg, fields...)
+}
+
+// Debugf logs a message at Debug level with a template.
+func (l *Logger) Debugf(template string, args ...any) {
+	l.sugaredLogger.Debugf(template, args...)
+}
+
+// Infof logs a message at Info level with a template.
+func (l *Logger) Infof(template string, args ...any) {
```
Comment (vladpistun) [inline pkg/logger/logger.go]: u really need all this staff with "f" suffix?

File: internal/storage/gormstorage/driver.go [hunk 17]
```diff
@@ -14,40 +18,44 @@ const RequestTimeout = time.Second * 5
...
-func (c *Connection) Setup(dsn string) error {
+// Setup sets up a new Connection with a logger.
+func (c *Connection) Setup(dsn string, l *logger.Logger) error {
+	c.l = l
 	var counts int64
 	for {
 		db, err := openDB(dsn)
 		if err != nil {
-			log.Printf("Postgres not yet ready... Attempt: %d\n", counts)
+			c.l.Error("Postgres not yet ready...", zap.Int64("attempt", counts), zap.Error(err))
```
Comment (agritsyk) [inline internal/storage/gormstorage/driver.go:36]: Тут точно error? Якщо ми ще чекаємо, що база підніметься, то можливо warning чи debug? А далі, якщо не піднялась (45 рядок в цьому ж файлі), то тоді error?

Comment (vladyslavpavlenko) [reply internal/storage/gormstorage/driver.go:36]: Першочергово використовував Warn, але @vladpistun порадив виводити логи про помилки на Error рівні 

https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-vladyslavpavlenko/pull/8#discussion_r1684888560

**Review by agritsyk — APPROVED** [review]

**Pull request merged**